JavaScript/chapter01/p08_zero_matrix/avc278.js

```diff
@@ -7,37 +7,46 @@ const assert = require("assert");
  * @param  {Array.<number[]>} matrix input 2d array to mutate
  * @return {null}
  *
- * Okay, so I have a jank solution that assumes no element in the array is currently -999.
- * We iterate through the array, and when we find a 0, we flag its row and column by updating all elements in there to
- * -999. The outer two for loop iterations is O(M * N) time, and each inner for loop is O(M) and O(N), respectively,
- * resulting in a grand total of O((M * N) * (M + N)) time complexity, while updating takes O(1) additional space.
- * When we take a second pass at the matrix, we update all elements that have been flagged -999 to 0, which is O(N^2)
- * time and O(1) additional space.
+ * We start by initializing rows and cols arrays with false values. We iterate through the matrix, and when we find a 0,
+ * we flag its row and column by updating the corresponding indices in rows and cols for future use.
+ * When we take a second pass at the matrix, we update all rows that have been flagged to 0, and all columns that have
+ * been flagged to 0, which is O(M + N).
  *
- * Runtime: O((M * N) * (M + N))
- * Space:   O(1)
+ * Runtime: O(M * N)
+ * Space:   O(M + N)
  *
  */
 const zeroMatrix = (arr) => {
+  if (arr.length === 0 || arr[0].length === 0) return;
+  const rows = [];
+  const cols = [];
+
+  for (let i = 0; i < arr.length; i++) {
+    rows.push(false);
+  }
+  for (let i = 0; i < arr[0].length; i++) {
+    cols.push(false);
+  }
```

Comment:
Using [`Array.fill()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fill) makes a bit more concise code:

```suggestion
  const rows = new Array(arr.length).fill(false);
  const cols = new Array(arr[0].length).fill(false);
```

Reply:
I'm a bit embarrassed to say I've never seen `Array.fill()` up until this point in my journey with JS. But yeah, you're totally right. This cleans it up by a lot.

Reply:
don't be embarrassed! https://xkcd.com/1053/

nobody is expected to know _the whole JavaScript API_. so all's good!